Zend/zend_inheritance.c

```diff
@@ -2243,21 +2281,28 @@ static bool do_trait_constant_check(zend_class_entry *ce, zend_class_constant *t
 
 	zend_class_constant *existing_constant = Z_PTR_P(zv);
 
-	if ((ZEND_CLASS_CONST_FLAGS(trait_constant) & flags_mask) != (ZEND_CLASS_CONST_FLAGS(existing_constant) & flags_mask) ||
-	    !check_trait_property_or_constant_value_compatibility(ce, &trait_constant->value, &existing_constant->value)) {
+	if ((ZEND_CLASS_CONST_FLAGS(trait_constant) & flags_mask) != (ZEND_CLASS_CONST_FLAGS(existing_constant) & flags_mask)) {
+		return emit_incompatible_trait_constant_error(ce, existing_constant, trait_constant, name, traits, current_trait);
+	}
+
+	if (ZEND_TYPE_IS_SET(trait_constant->type) != ZEND_TYPE_IS_SET(existing_constant->type)) {
+		return emit_incompatible_trait_constant_error(ce, existing_constant, trait_constant, name, traits, current_trait);
+	} else if (ZEND_TYPE_IS_SET(trait_constant->type)) {
+		inheritance_status status1 = zend_perform_covariant_type_check(ce, existing_constant->type, traits[current_trait], trait_constant->type);
+		inheritance_status status2 = zend_perform_covariant_type_check(traits[current_trait], trait_constant->type, ce, existing_constant->type);
+		if (status1 == INHERITANCE_ERROR || status2 == INHERITANCE_ERROR) {
+			return emit_incompatible_trait_constant_error(ce, existing_constant, trait_constant, name, traits, current_trait);
+		}
+	}
+
+	if (!check_trait_property_or_constant_value_compatibility(ce, &trait_constant->value, &existing_constant->value)) {
```

Comment:
Not your code, but it looks like `check_trait_property_or_constant_value_compatibility` is missing error handling. It doesn't check whether `zval_update_constant_ex` was successful.

Reply:
Nice catch again!